lib/auth/saml.go

```diff
@@ -434,36 +539,38 @@ func (a *Server) validateSAMLResponse(samlResponse string) (*samlAuthResponse, e
 			SessionTTL: params.sessionTTL,
 			LoginTime:  a.clock.Now().UTC(),
 		})
+
 		if err != nil {
-			return re, trace.Wrap(err)
+			return nil, trace.Wrap(err).AddUserMessage("Failed to create web session.")
 		}
 
-		re.auth.Session = session
+		auth.Session = session
 	}
 
 	// If a public key was provided, sign it and return a certificate.
 	if len(request.PublicKey) != 0 {
 		sshCert, tlsCert, err := a.createSessionCert(user, params.sessionTTL, request.PublicKey, request.Compatibility, request.RouteToCluster, request.KubernetesCluster)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return nil, trace.Wrap(err).AddUserMessage("Failed to create session certificate.")
 		}
 		clusterName, err := a.GetClusterName()
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return nil, trace.Wrap(err).AddUserMessage("Failed to obtain cluster name.")
 		}
-		re.auth.Cert = sshCert
-		re.auth.TLSCert = tlsCert
+		auth.Cert = sshCert
+		auth.TLSCert = tlsCert
 
 		// Return the host CA for this cluster only.
 		authority, err := a.GetCertAuthority(ctx, types.CertAuthID{
 			Type:       types.HostCA,
 			DomainName: clusterName.GetClusterName(),
 		}, false)
 		if err != nil {
-			return nil, trace.Wrap(err)
+			return nil, trace.Wrap(err).AddUserMessage("Failed to obtain cluster's host CA.")
 		}
-		re.auth.HostSigners = append(re.auth.HostSigners, authority)
+		auth.HostSigners = append(auth.HostSigners, authority)
 	}
 
-	return re, nil
+	diagCtx.info.Success = "regular flow"
```

Comment:
Following up on the comment above, looks like this is how you differentiate b/w "test" and "regular" flow. Maybe rather make a constant like `SSOFlowType` and use it in the diag context?

```go
type SSOFlowType string
var (
  SSORegularFlow = "regular"
  SSOTestFlow = "test"
)
```

Reply:
I think I'll make it simpler still: add bool `TestFlow` to `SSODiagnosticInfo` and make `Success` into `bool` too.